docs/coding-guidelines.md

```diff
@@ -0,0 +1,269 @@
+    match self.child_1.poll(cx) {
+      // The child made progress.
+      Poll::Ready(_) => {
+        // Either return an event to the parent:
+        return Poll::Ready(todo!());
+        // or `continue`, thus polling `child_1` again. `child_1` can potentially make more progress:
+        continue
+        // but NEVER move to the next child if the current child made progress. Given
+        // that the current child might be able to make more progress, it did not yet
+        // register the waker in order for the root task to be woken up later on. Moving
+        // on to the next child might result in the larger `Future` task to stall as it
+        // assumes that there is no more progress to be made.
+      }
+
+      // The child did not make progress. It has registered the waker for a
+      // later wake up. Proceed with the other childs.
+      Poll::Ready(_) => {
+    }
+
+    match self.child_2.poll(cx) {
+      // As above.
+    }
+
+    match self.child_3.poll(cx) {
+      // As above.
+    }
+
+    // ...
+
+    // None of the child state machines can make any more progress. Each registered
+    // the waker in order for the root `Future` task to be woken up again.
+    return Poll::Pending
+  }
+}
+```
+
+### Prioritize local work over new work from a remote
+
+When handling multiple work streams, prioritize local work items over
+accepting new work items from a remote. Take the following state machine as an
+example, reading and writing from a socket, returning result to its parent:
+
+``` rust
+struct SomeStateMachine {
+  socket: Socket,
+  events_to_return_to_parent: VecDeque<Event>,
+  messages_to_send_on_socket: VecDeque<Message>,
+}
+
+impl Stream for SomeStateMachine {
+  type Item = Event;
+
+  fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+    // First priority is returning local finished work.
+    if let Some(event) = events_to_return_to_parent.pop_front() {
+      return Poll::Ready(Some(event));
+    }
+
+    // Second priority is finishing local work, i.e. sending on the socket.
+    if let Poll::Ready(()) = socket.poll_ready(cx) {
+      todo!("Send messages")
+    }
+
+    // Last priority is accepting new work, i.e. reading from the socket.
+    if let Some(work_item) = socket.poll_next(cx) {
+      todo!("Start work on new work item")
+    }
+  }
+}
+```
+
+This priotization provides:
+- Low memory footprint as local queues (here `events_to_return_to_parent`) stay small.
+- Low latency as accepted local work is not stuck in queues.
+- DOS defense as a remote does not control the size of the local queue, nor starves local work with its remote work.
+
+## Bound everything
+
+The concept of unboundedness is an illusion. Use bounded mechanisms to prevent
+unbounded memory growth and high latencies.
+
+### Channels
+
+When using channels (e.g. `futures::channel::mpsc` or `std::sync::mpsc`)
+always use the bounded variant, never use the unbounded variant. When using a
```

Comment:
We should use clippy's mechanism of banning these functions!

https://rust-lang.github.io/rust-clippy/master/index.html#disallowed_methods

Reply:
Good idea. Though I am failing to have clippy pick up my `Clippy.toml` when I save it in the root of the repository.

``` toml
disallowed-methods = [
  { path = "futures::sync::mpsc::channel", reason = "does not enforce backpressure" },
]
```

@thomaseizinger does this work for you?

Reply:
Will be introduced with https://github.com/libp2p/rust-libp2p/pull/2823.